internal/modelcontroller/cache.go

```diff
@@ -0,0 +1,469 @@
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("parsing pvc model annotation: %w", err)
+	}
+
+	// Run Job to populate PVC if not already downloaded.
+	if pvcModelAnn.UID != string(model.UID) {
+		// Ensure the download job exists.
+		if !jobExists {
+			job = r.loadCacheJobForModel(model, cfg)
+			if err := ctrl.SetControllerReference(model, job, r.Scheme); err != nil {
+				return ctrl.Result{}, fmt.Errorf("setting controller reference on job: %w", err)
+			}
+			if err := r.Create(ctx, job); err != nil {
+				return ctrl.Result{}, fmt.Errorf("creating job: %w", err)
+			}
+			return ctrl.Result{}, errReturnEarly
+		}
+
+		if !k8sutils.IsJobCompleted(job) {
+			return ctrl.Result{}, errReturnEarly
+		}
+		if err := r.updatePVCModelAnnotation(ctx, pvc, model.Name, PVCModelAnnotationValue{
+			UID:       string(model.UID),
+			Timestamp: time.Now(),
+		}); err != nil {
+			return ctrl.Result{}, fmt.Errorf("setting pvc model annotation: %w", err)
+		}
+	}
+
+	if jobExists {
+		// Delete Job.
+		// Use foreground deletion policy to ensure the Pods are deleted as well.
+		if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("deleting job: %w", err)
+		}
+	}
+
+	model.Status.Cache.Loaded = pvcModelAnn.UID == string(model.UID)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ModelReconciler) finalizeCache(ctx context.Context, model *kubeaiv1.Model, cfg ModelConfig) error {
+	pvc := &corev1.PersistentVolumeClaim{}
+	var pvcNotFound bool
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: model.Namespace,
+		Name:      cachePVCName(model, cfg),
+	}, pvc); err != nil {
+		if apierrors.IsNotFound(err) {
+			pvcNotFound = true
+		} else {
+			return fmt.Errorf("getting cache PVC: %w", err)
+		}
+	}
+
+	if pvcNotFound || pvc.DeletionTimestamp != nil {
+		// If the PVC is not found or is already being deleted, delete all cache jobs and pods.
+		// No need trying to update the PVC annotations or perform other cleanup.
+		if err := r.deleteAllCacheJobsAndPods(ctx, model); err != nil {
+			return fmt.Errorf("deleting all cache jobs and pods: %w", err)
+		}
+		if controllerutil.RemoveFinalizer(model, kubeaiv1.ModelCacheEvictionFinalizer) {
+			if err := r.Update(ctx, model); err != nil {
+				return fmt.Errorf("removing cache deletion finalizer: %w", err)
+			}
+		}
+		return nil
+	}
+
+	if controllerutil.ContainsFinalizer(model, kubeaiv1.ModelCacheEvictionFinalizer) {
+		job := &batchv1.Job{}
+		var jobExists bool
+		if err := r.Client.Get(ctx, types.NamespacedName{
+			Namespace: model.Namespace,
+			Name:      evictCacheJobName(model),
+		}, job); err != nil {
+			if apierrors.IsNotFound(err) {
+				jobExists = false
+			} else {
+				return fmt.Errorf("getting cache deletion job: %w", err)
+			}
+		} else {
+			jobExists = true
+		}
+
+		if !jobExists {
+			job := r.evictCacheJobForModel(model, cfg)
```

Comment:
nit: this shadows  the var in L165. Let's reuse it instead
```suggestion
			job = r.evictCacheJobForModel(model, cfg)
```

Reply:
Renamed Job vars to be more explicit on what they are